EVCC crash on unconfigured OCPP charger (Denial of Service)

### Describe the bug

I have two ABB Terra AC W22-T-0, one is configured in EVCC the other is not. The configured one connects without an issue. As soon as the second (unconfigured) charger connects - EVCC will `SIGSEGV`.

EVCC enters a crash-loop, as soon as it's back online and the second charger connects it will crash again.

### Steps to reproduce

1. Let an unconfigured OCPP enabled charger connect to EVCC
2. EVCC will crash

### Configuration details

```yaml
chargers:
  - type: template
    template: ocpp
    stationid: TACW2243521T1625
    connector: 1
    idtag: 1
    connecttimeout: 15m
    timeout: 5m
    getconfiguration: false
    bootnotification: true
    meterinterval: 1m
    metervalues: 1m
    name: wallbox-2
```

### Log details

```text
Using config file: /config/evcc.yaml
starting evcc: 'EVCC_DATABASE_DSN=/data/evcc.db evcc --config /config/evcc.yaml'
[main  ] INFO 2023/09/24 08:36:10 evcc 0.120.2
[main  ] INFO 2023/09/24 08:36:10 using config file: /config/evcc.yaml
[main  ] INFO 2023/09/24 08:36:10 starting ui and api at :7070
[TACW2243521T1625] DEBUG 2023/09/24 08:36:11 waiting for chargepoint: 15m0s
[ocpp  ] DEBUG 2023/09/24 08:38:42 chargepoint connected: TACW2243521T1625
[ocpp  ] WARN 2023/09/24 08:38:42 chargepoint connected, unknown: TACW2243521T2288
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x175d412]
goroutine 31 [running]:
github.com/evcc-io/evcc/charger/ocpp.(*CP).BootNotification(0x40feda?, 0xc0009d0005?)
	github.com/evcc-io/evcc/charger/ocpp/cp_core.go:29 +0x12
github.com/evcc-io/evcc/charger/ocpp.(*CS).OnBootNotification(0xc00096cbd0?, {0xc0009d0005?, 0x10?}, 0x32f2ba0?)
	github.com/evcc-io/evcc/charger/ocpp/cs_core.go:76 +0xee
github.com/lorenzodonini/ocpp-go/ocpp1%2e6.(*centralSystem).handleIncomingRequest.func1()
	github.com/lorenzodonini/ocpp-go@v0.17.0/ocpp1.6/central_system.go:508 +0x4ea
created by github.com/lorenzodonini/ocpp-go/ocpp1%2e6.(*centralSystem).handleIncomingRequest in goroutine 48
	github.com/lorenzodonini/ocpp-go@v0.17.0/ocpp1.6/central_system.go:505 +0x5ac
```


### What type of operating system are you running?

HomeAssistant Add-on

### Version

0.120.2

--- a/charger/ocpp/const.go
+++ b/charger/ocpp/const.go
@@ -1,0 +1,20 @@
+package ocpp
+
+const (
+	// Core profile keys
+	KeyNumberOfConnectors = "NumberOfConnectors"
+
+	// Meter profile keys
+	KeyMeterValuesSampledData   = "MeterValuesSampledData"
+	KeyMeterValueSampleInterval = "MeterValueSampleInterval"
+
+	// Smart Charging profile keys
+	KeyChargeProfileMaxStackLevel              = "ChargeProfileMaxStackLevel"
+	KeyChargingScheduleAllowedChargingRateUnit = "ChargingScheduleAllowedChargingRateUnit"
+	KeyChargingScheduleMaxPeriods              = "ChargingScheduleMaxPeriods"
+	KeyConnectorSwitch3to1PhaseSupported       = "ConnectorSwitch3to1PhaseSupported"
+	KeyMaxChargingProfilesInstalled            = "MaxChargingProfilesInstalled"
+
+	// Alfen specific keys
+	KeyAlfenPlugAndChargeIdentifier = "PlugAndChargeIdentifier"
+)

--- a/charger/ocpp/cp.go
+++ b/charger/ocpp/cp.go
@@ -15,25 +15,6 @@ import (
 	"github.com/lorenzodonini/ocpp-go/ocpp1.6/types"
 )
 
-const (
-	// Core profile keys
-	KeyNumberOfConnectors = "NumberOfConnectors"
-
-	// Meter profile keys
-	KeyMeterValuesSampledData   = "MeterValuesSampledData"
-	KeyMeterValueSampleInterval = "MeterValueSampleInterval"
-
-	// Smart Charging profile keys
-	KeyChargeProfileMaxStackLevel              = "ChargeProfileMaxStackLevel"
-	KeyChargingScheduleAllowedChargingRateUnit = "ChargingScheduleAllowedChargingRateUnit"
-	KeyChargingScheduleMaxPeriods              = "ChargingScheduleMaxPeriods"
-	KeyConnectorSwitch3to1PhaseSupported       = "ConnectorSwitch3to1PhaseSupported"
-	KeyMaxChargingProfilesInstalled            = "MaxChargingProfilesInstalled"
-
-	// Alfen specific keys
-	KeyAlfenPlugAndChargeIdentifier = "PlugAndChargeIdentifier"
-)
-
 // TODO support multiple connectors
 // Since ocpp-go interfaces at charge point level, we need to manage multiple connector separately
 

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -66,28 +66,7 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
 
-	if !cs.remoteByID(chargePoint.ID()) {
-		// check for anonymous chargepoint
-		if cp, err := cs.chargepointByID(""); err == nil {
-			cs.log.INFO.Printf("charge point connected, registering: %s", chargePoint.ID())
-
-			// update id
-			cp.RegisterID(chargePoint.ID())
-
-			cs.cps[chargePoint.ID()] = cp
-			delete(cs.cps, "")
-
-			cp.connect(true)
-
-			return
-		}
-
-		cs.log.WARN.Printf("charge point connected, unknown: %s", chargePoint.ID())
-
-		// register unknown chargepoint
-		// when chargepoint setup is complete, it will eventually be associated with the connected id
-		cs.cps[chargePoint.ID()] = nil
-	} else {
+	if cs.remoteByID(chargePoint.ID()) {
 		cs.log.DEBUG.Printf("charge point connected: %s", chargePoint.ID())
 
 		// trigger initial connection if chargepoint is already setup
@@ -95,6 +74,27 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 			cp.connect(true)
 		}
 	}
+
+	// check for anonymous chargepoint
+	if cp, err := cs.chargepointByID(""); err == nil {
+		cs.log.INFO.Printf("charge point connected, registering: %s", chargePoint.ID())
+
+		// update id
+		cp.RegisterID(chargePoint.ID())
+
+		cs.cps[chargePoint.ID()] = cp
+		delete(cs.cps, "")
+
+		cp.connect(true)
+
+		return
+	}
+
+	cs.log.WARN.Printf("charge point connected, unknown: %s", chargePoint.ID())
+
+	// register unknown chargepoint
+	// when chargepoint setup is complete, it will eventually be associated with the connected id
+	cs.cps[chargePoint.ID()] = nil
 }
 
 // ChargePointDisconnected implements ocpp16.ChargePointConnectionHandler
@@ -102,16 +102,9 @@ func (cs *CS) ChargePointDisconnected(chargePoint ocpp16.ChargePointConnection) 
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
 
-	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
-		cs.log.ERROR.Printf("chargepoint disconnected: %v", err)
-	} else {
-		cs.log.DEBUG.Printf("chargepoint disconnected: %s", chargePoint.ID())
+	cs.log.DEBUG.Printf("charge point disconnected: %s", chargePoint.ID())
 
-		if cp == nil {
-			// remove unknown chargepoint
-			delete(cs.cps, chargePoint.ID())
-		} else {
-			cp.connect(false)
-		}
+	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
+		cp.connect(false)
 	}
 }

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -73,6 +73,8 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 		if cp, _ := cs.chargepointByID(chargePoint.ID()); cp != nil {
 			cp.connect(true)
 		}
+
+		return
 	}
 
 	// check for anonymous charge point

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -89,7 +89,7 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 		return
 	}
 
-	cs.log.WARN.Printf("charge point connected, unknown: %s", chargePoint.ID())
+	cs.log.WARN.Printf("unknown charge point connected: %s", chargePoint.ID())
 
 	// register unknown charge point
 	// when charge point setup is complete, it will eventually be associated with the connected id

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -16,17 +16,17 @@ type CS struct {
 	cps map[string]*CP
 }
 
-// Register registers a chargepoint with the central system.
-// The chargepoint identified by id may already be connected in which case initial connection is triggered.
+// Register registers a charge point with the central system.
+// The charge point identified by id may already be connected in which case initial connection is triggered.
 func (cs *CS) Register(id string, cp *CP) error {
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
 
 	if _, ok := cs.cps[id]; ok && id == "" {
-		return errors.New("cannot have >1 chargepoint with empty station id")
+		return errors.New("cannot have >1 charge point with empty station id")
 	}
 
-	// trigger unknown chargepoint connected
+	// trigger unknown charge point connected
 	if unknown, ok := cs.cps[id]; ok && unknown == nil {
 		cp.connect(true)
 	}
@@ -43,13 +43,13 @@ func (cs *CS) errorHandler(errC <-chan error) {
 	}
 }
 
-// remoteByID returns a connected remote chargepoint identified by id.
+// remoteByID returns a connected remote charge point identified by id.
 func (cs *CS) remoteByID(id string) bool {
 	_, ok := cs.cps[id]
 	return ok
 }
 
-// chargepointByID returns a configured chargepoint identified by id.
+// chargepointByID returns a configured charge point identified by id.
 func (cs *CS) chargepointByID(id string) (*CP, error) {
 	cp, ok := cs.cps[id]
 	if !ok {
@@ -69,13 +69,13 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 	if cs.remoteByID(chargePoint.ID()) {
 		cs.log.DEBUG.Printf("charge point connected: %s", chargePoint.ID())
 
-		// trigger initial connection if chargepoint is already setup
+		// trigger initial connection if charge point is already setup
 		if cp, _ := cs.chargepointByID(chargePoint.ID()); cp != nil {
 			cp.connect(true)
 		}
 	}
 
-	// check for anonymous chargepoint
+	// check for anonymous charge point
 	if cp, err := cs.chargepointByID(""); err == nil {
 		cs.log.INFO.Printf("charge point connected, registering: %s", chargePoint.ID())
 
@@ -92,8 +92,8 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 
 	cs.log.WARN.Printf("charge point connected, unknown: %s", chargePoint.ID())
 
-	// register unknown chargepoint
-	// when chargepoint setup is complete, it will eventually be associated with the connected id
+	// register unknown charge point
+	// when charge point setup is complete, it will eventually be associated with the connected id
 	cs.cps[chargePoint.ID()] = nil
 }
 

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -43,12 +43,6 @@ func (cs *CS) errorHandler(errC <-chan error) {
 	}
 }
 
-// remoteByID returns a connected remote charge point identified by id.
-func (cs *CS) remoteByID(id string) bool {
-	_, ok := cs.cps[id]
-	return ok
-}
-
 // chargepointByID returns a configured charge point identified by id.
 func (cs *CS) chargepointByID(id string) (*CP, error) {
 	cp, ok := cs.cps[id]
@@ -66,19 +60,22 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
 
-	if cs.remoteByID(chargePoint.ID()) {
+	// check for configured charge point
+	cp, ok := cs.cps[chargePoint.ID()]
+	if ok {
 		cs.log.DEBUG.Printf("charge point connected: %s", chargePoint.ID())
 
 		// trigger initial connection if charge point is already setup
-		if cp, _ := cs.chargepointByID(chargePoint.ID()); cp != nil {
+		if cp != nil {
 			cp.connect(true)
 		}
 
 		return
 	}
 
-	// check for anonymous charge point
-	if cp, err := cs.chargepointByID(""); err == nil {
+	// check for configured anonymous charge point
+	cp, ok = cs.cps[""]
+	if ok && cp != nil {
 		cs.log.INFO.Printf("charge point connected, registering: %s", chargePoint.ID())
 
 		// update id

--- a/charger/ocpp/cs.go
+++ b/charger/ocpp/cs.go
@@ -43,22 +43,33 @@ func (cs *CS) errorHandler(errC <-chan error) {
 	}
 }
 
+// remoteByID returns a connected remote chargepoint identified by id.
+func (cs *CS) remoteByID(id string) bool {
+	_, ok := cs.cps[id]
+	return ok
+}
+
+// chargepointByID returns a configured chargepoint identified by id.
 func (cs *CS) chargepointByID(id string) (*CP, error) {
 	cp, ok := cs.cps[id]
 	if !ok {
 		return nil, fmt.Errorf("unknown charge point: %s", id)
 	}
+	if cp == nil {
+		return nil, fmt.Errorf("charge point not configured: %s", id)
+	}
 	return cp, nil
 }
 
+// NewChargePoint implements ocpp16.ChargePointConnectionHandler
 func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 	cs.mu.Lock()
 	defer cs.mu.Unlock()
 
-	if cp, err := cs.chargepointByID(chargePoint.ID()); err != nil {
+	if !cs.remoteByID(chargePoint.ID()) {
 		// check for anonymous chargepoint
 		if cp, err := cs.chargepointByID(""); err == nil {
-			cs.log.INFO.Printf("chargepoint connected, registering: %s", chargePoint.ID())
+			cs.log.INFO.Printf("charge point connected, registering: %s", chargePoint.ID())
 
 			// update id
 			cp.RegisterID(chargePoint.ID())
@@ -71,21 +82,22 @@ func (cs *CS) NewChargePoint(chargePoint ocpp16.ChargePointConnection) {
 			return
 		}
 
-		cs.log.WARN.Printf("chargepoint connected, unknown: %s", chargePoint.ID())
+		cs.log.WARN.Printf("charge point connected, unknown: %s", chargePoint.ID())
 
 		// register unknown chargepoint
 		// when chargepoint setup is complete, it will eventually be associated with the connected id
 		cs.cps[chargePoint.ID()] = nil
 	} else {
-		cs.log.DEBUG.Printf("chargepoint connected: %s", chargePoint.ID())
+		cs.log.DEBUG.Printf("charge point connected: %s", chargePoint.ID())
 
 		// trigger initial connection if chargepoint is already setup
-		if cp != nil {
+		if cp, _ := cs.chargepointByID(chargePoint.ID()); cp != nil {
 			cp.connect(true)
 		}
 	}
 }
 
+// ChargePointDisconnected implements ocpp16.ChargePointConnectionHandler
 func (cs *CS) ChargePointDisconnected(chargePoint ocpp16.ChargePointConnection) {
 	cs.mu.Lock()
 	defer cs.mu.Unlock()

--- a/charger/ocpp/cs_core.go
+++ b/charger/ocpp/cs_core.go
@@ -1,6 +1,8 @@
 package ocpp
 
 import (
+	"fmt"
+
 	"github.com/lorenzodonini/ocpp-go/ocpp1.6/core"
 	"github.com/lorenzodonini/ocpp-go/ocpp1.6/firmware"
 	"github.com/lorenzodonini/ocpp-go/ocpp1.6/remotetrigger"
@@ -73,6 +75,7 @@ func (cs *CS) OnBootNotification(id string, request *core.BootNotificationReques
 		return nil, err
 	}
 
+	fmt.Println("OnBootNotification id", id)
 	return cp.BootNotification(request)
 }
 

--- a/charger/ocpp_test.go
+++ b/charger/ocpp_test.go
@@ -27,50 +27,52 @@ func TestOcpp(t *testing.T) {
 type ocppTestSuite struct {
 	suite.Suite
 	clock *clock.Mock
-	cp    ocpp16.ChargePoint
 }
 
 func (suite *ocppTestSuite) SetupSuite() {
-	// setup cs
-	suite.NotNil(ocpp.Instance())
-
-	// setup cp
 	suite.clock = clock.NewMock()
-	cp := ocpp16.NewChargePoint("test", nil, nil)
+	suite.NotNil(ocpp.Instance())
+}
 
+func (suite *ocppTestSuite) startChargePoint(id string) ocpp16.ChargePoint {
 	// set a handler for all callback functions
-	triggerC := make(chan remotetrigger.MessageTrigger, 1)
-	handler := &ChargePointHandler{triggerC: triggerC}
+	handler := &ChargePointHandler{
+		triggerC: make(chan remotetrigger.MessageTrigger, 1),
+	}
+
+	// create charge point with handler
+	cp := ocpp16.NewChargePoint(id, nil, nil)
 	cp.SetCoreHandler(handler)
 	cp.SetRemoteTriggerHandler(handler)
 
+	// let cs handle the trigger messages
 	go func() {
-		for msg := range triggerC {
-			suite.handleTrigger(msg)
+		for msg := range handler.triggerC {
+			suite.handleTrigger(cp, msg)
 		}
 	}()
 
-	suite.cp = cp
+	return cp
 }
 
-func (suite *ocppTestSuite) handleTrigger(msg remotetrigger.MessageTrigger) {
+func (suite *ocppTestSuite) handleTrigger(cp ocpp16.ChargePoint, msg remotetrigger.MessageTrigger) {
 	switch msg {
 	case core.BootNotificationFeatureName:
-		if res, err := suite.cp.BootNotification("demo", "evcc"); err != nil {
+		if res, err := cp.BootNotification("demo", "evcc"); err != nil {
 			suite.T().Log("BootNotification:", err)
 		} else {
 			suite.T().Log("BootNotification:", res)
 		}
 
 	case core.StatusNotificationFeatureName:
-		if res, err := suite.cp.StatusNotification(ocppTestConnector, core.NoError, core.ChargePointStatusAvailable); err != nil {
+		if res, err := cp.StatusNotification(ocppTestConnector, core.NoError, core.ChargePointStatusAvailable); err != nil {
 			suite.T().Log("StatusNotification:", err)
 		} else {
 			suite.T().Log("StatusNotification:", res)
 		}
 
 	case core.MeterValuesFeatureName:
-		if res, err := suite.cp.MeterValues(1, []types.MeterValue{
+		if res, err := cp.MeterValues(1, []types.MeterValue{
 			{
 				Timestamp: types.NewDateTime(suite.clock.Now()),
 				SampledValue: []types.SampledValue{
@@ -91,20 +93,21 @@ func (suite *ocppTestSuite) handleTrigger(msg remotetrigger.MessageTrigger) {
 
 func (suite *ocppTestSuite) TestConnect() {
 	// start cp client
-	suite.NoError(suite.cp.Start(ocppTestUrl))
-	suite.True(suite.cp.IsConnected())
+	cp := suite.startChargePoint("test")
+	suite.NoError(cp.Start(ocppTestUrl))
+	suite.True(cp.IsConnected())
 
 	// start cp server
 	c, err := NewOCPP("test", ocppTestConnector, "", "", 0, false, false, ocppTestConnectTimeout, ocppTestTimeout)
-	suite.NoError(err)
-
 	if err != nil {
+		suite.NoError(err)
 		return
 	}
 
 	suite.clock.Add(ocppTestTimeout)
 	c.cp.TestClock(suite.clock)
 
+	// status
 	_, err = c.Status()
 	suite.NoError(err)
 
@@ -117,4 +120,12 @@ func (suite *ocppTestSuite) TestConnect() {
 	f, err = c.totalEnergy()
 	suite.NoError(err)
 	suite.Equal(1.2, f)
+
+	// 2nd charge point
+	cp2 := suite.startChargePoint("test2")
+	suite.NoError(cp2.Start(ocppTestUrl))
+	suite.True(cp2.IsConnected())
+
+	_, err = cp2.BootNotification("demo", "evcc")
+	suite.NoError(err)
 }

--- a/charger/ocpp_test.go
+++ b/charger/ocpp_test.go
@@ -126,6 +126,7 @@ func (suite *ocppTestSuite) TestConnect() {
 	suite.NoError(cp2.Start(ocppTestUrl))
 	suite.True(cp2.IsConnected())
 
+	// error on unconfigured 2nd charge point
 	_, err = cp2.BootNotification("demo", "evcc")
-	suite.NoError(err)
+	suite.Error(err)
 }

